library/nist_kw.c

```diff
@@ -0,0 +1,650 @@
+         * for KW is 3 to 2^54 semiblocks
+         */
+        if( ( in_len < 24 )
+#if SIZE_MAX > UINT_MAX
+            || ( in_len > 0x200000000000000 )
+#endif
+            )
+        {
+            return ( MBEDTLS_ERR_CIPHER_BAD_INPUT_DATA );
+        }
+
+        ret = unwrap( ctx, input, in_len / KW_SEMIBLOCK_LENGTH, output, out_len);
+        if( ret != 0 )
+            return ( ret );
+
+        if ( memcmp( ICV1, output, KW_SEMIBLOCK_LENGTH ) != 0 )
+        {
+            ret = MBEDTLS_ERR_CIPHER_AUTH_FAILED;
+            goto cleanup;
+        }
+
+        /*
+         * shift output to point to P
+         */
+        for( i=0; i < ( *out_len / KW_SEMIBLOCK_LENGTH );i++ )
+        {
+            memcpy( output + (i * KW_SEMIBLOCK_LENGTH), output + ( ( i + 1 ) * KW_SEMIBLOCK_LENGTH), KW_SEMIBLOCK_LENGTH );
+        }
+        *out_len = *out_len - KW_SEMIBLOCK_LENGTH;
+    }
+    else // KWP
+    {
+        int padlen = 0;
+        uint32_t Plen;
+        /*
+         * According to SP800-38F Table 1, ciphertext length size limits
+         * for KWP is 2 to 2^29 semiblocks
+         */
+        if( ( in_len < KW_SEMIBLOCK_LENGTH * 2 )
+#if SIZE_MAX > UINT_MAX
+         || ( in_len > 0x100000000 )
+#endif
+        )
+        {
+            return ( MBEDTLS_ERR_CIPHER_BAD_INPUT_DATA );
+        }
+
+        if( in_len == KW_SEMIBLOCK_LENGTH * 2 )
+        {
+            ret = mbedtls_cipher_update( &ctx->cipher_ctx, input, 16, output, out_len);
+            if( ret != 0 )
+                goto cleanup;
+        }
+        else if( in_len >  KW_SEMIBLOCK_LENGTH * 2 )
+        {
+            ret = unwrap( ctx, input, in_len / KW_SEMIBLOCK_LENGTH, output, out_len);
+            if( ret != 0 )
+                return ( ret );
+        }
+        if( memcmp( ICV2, output, KW_SEMIBLOCK_LENGTH / 2 ) != 0 )
+        {
+            ret = MBEDTLS_ERR_CIPHER_AUTH_FAILED;
+            goto cleanup;
+        }
+
+        GET_UINT32_BE( Plen, output, KW_SEMIBLOCK_LENGTH / 2 );
+
+        padlen = 8 * ( ( in_len / 8 ) - 1 ) - Plen;
+        if ( padlen < 0 || padlen > 7 )
+        {
+            ret = MBEDTLS_ERR_CIPHER_AUTH_FAILED;
+            goto cleanup;
+        }
+        for( i = 0; i < (size_t)padlen; i++ )
+        {
+            if( output[ in_len - padlen + i ] != 0 )
+            {
+                ret = MBEDTLS_ERR_CIPHER_AUTH_FAILED;
+                goto cleanup;
+            }
+        }
+        /*
+         * shift output to point to P
+         */
+
+        memcpy( output, output + KW_SEMIBLOCK_LENGTH , Plen);
+        memset( output + Plen, 0, KW_SEMIBLOCK_LENGTH + padlen );
+        *out_len = Plen;
+    }
+
+cleanup:
+    if( ret != 0 )
+        mbedtls_platform_zeroize( output, *out_len - 1);
+
+    mbedtls_cipher_finish( &ctx->cipher_ctx, NULL, &i );
+    return( ret );
+}
+
+#endif /* !MBEDTLS_NIST_KW_ALT */
+
+#if defined(MBEDTLS_SELF_TEST) && defined(MBEDTLS_AES_C)
+
+#define KW_TESTS 3
```

Comment:
Not a blocker at all, but consider removing this define and using `[]` and a constant with the value of`sizeof( key_len ) / sizeof( *key_len )` instead.

Reply:
ok, I actually copied this from the CCM tests. I wanted to be consistent

Reply:
I would say it's more readable with this define.
There are just too many arrays using this value in this test

Reply:
OK, I am fine with this.